### ISSUE TITLE
Node.removeFromParent does not set parent of child node to null

### ISSUE THREAD

**Issue by roeltje25** [vaadin/flow#8586]
### Description of the bug
While debugging another issue for Vaadin CDI (see PR vaadin/cdi#344 and issue vaadin/cdi#345). I found that when removing a node from its parent will not remove the parent from the child. So consider a `Component c` with an `Element e = c.getElement()` and a parent `Element p = e.getParent()`. Now consider I would remove the element `e` from its parent `p` by `e.removeFromParent()` now the parent `p` will not contain `e` anymore, i.e. the following is `true`
```
p.getChildren().noneMatch(n -> n == e)
```
However, I can still do `e.getParent()`, which will result in `p`, i.e. the following is `false`:
```
e.getParent() == null
```
### Minimal reproducible example
```java
removeFromParentAndCheck(HasElement he) {
  Element e = he.getElement();
  Element p = e.getParent();
  e.removeFromParent();
  assert p.getChildren().noneMatch(n -> n == e) : "element e should be removed";
  assert e.getParent() == null : "e should now have no parent";
}
```
### Expected behavior
I basically would expect the child to have *no* parent
### Actual behavior
I observe that the child still has his old parent
### Versions:
    - Vaadin / Flow version: 14.2.0
    - Java version: JDK 11.0.7
    - OS version: Windows 10

**Comment by Legioth** [comment]
You're not showing how `he` is created, and I cannot reproduce it with a naive setup implementation. 

Another observation is that `Element` instances are flyweights that get recreated on demand, which means that you need to use `equals` instead of `==` in the `noneMatch` filter, thus making your current test a no-op. Maybe your example has a detach listener that puts the child back or something like that?

What I've tested is to add this as a simple test case in `com.vaadin.flow.dom.ElementTest`
```
    @Test
    public void removeFromParent_removesFromParent() {
        Element p = new Element("parent");
        Element e = new Element("child");
        p.appendChild(e);

        // Sanity checks
        Assert.assertTrue(p.getChildren().anyMatch(n -> n.equals(e)));
        Assert.assertEquals(p, e.getParent());

        e.removeFromParent();
        Assert.assertTrue("element e should be removed",
                p.getChildren().noneMatch(n -> n.equals(e)));

        Assert.assertNull("e should now have no parent", e.getParent());
    }
```

**Comment by roeltje25** [comment]
Good comment about the test. I just wrote it in this post and did not actually run that test.
The reason I walked into this issue is that We were having an issue with Vaadin CDI (see PR vaadin/cdi#344 and issue vaadin/cdi#345) where a nested layout component was attached multiple times after navigating away and clicking back. In order to fix it we added code in the routerlayout to remove the component on detach.

After investigating the problem I suggested a fix that would solve the problem during the handling of @RouteScoped when the scope ended and the component should be destroyed (and then running removeFromParent). However, our old code was also executed still, resulting in a double call to removeFromParent. As the parent is not removed from the child (only the other way around). the second removeFromParent would then fail with an exception telling that the child was not present on the parent.

Another test could therefore be the following (which indicates the same underlying issue I think).
```
    @Test
    public void removeFromParent_removesFromParent() {
        Element p = new Element("parent");
        Element e = new Element("child");
        p.appendChild(e);

        e.removeFromParent();
        // Following line should IMHO not throw an exception
        e.removeFromParent();
    }
```

**Comment by roeltje25** [comment]
I find it also strange that this is actually tested in https://github.com/vaadin/flow/blob/master/flow-server/src/test/java/com/vaadin/flow/dom/AbstractNodeTest.java
on line 525:
```
    @Test
    public void removeFromParent() {
        Node<?> parent = createParentNode();
        Element otherElement = new Element("other");
        parent.appendChild(otherElement);
        Assert.assertEquals(parent, otherElement.getParentNode());
        otherElement.removeFromParent();
        Assert.assertNull(otherElement.getParentNode());
    }
```

I am going to see if that test runs well for me and get back to you

**Comment by roeltje25** [comment]
The tests work out fine. So I am now investigating why it is a problem in my context. Will post my results later today

**Comment by roeltje25** [comment]
In our application it still does not work and gives this exception:
```
java.lang.IllegalArgumentException: Trying to detach an element from parent that does not have it.
	at com.vaadin.flow.dom.impl.AbstractNodeStateProvider.removeChild(AbstractNodeStateProvider.java:123) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.dom.Node.removeChild(Node.java:376) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.dom.Element.removeFromParent(Element.java:577) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.router.RouterLayout.removeRouterLayoutContent(RouterLayout.java:67) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.cdi.context.AbstractContextualStorageManager.lambda$7(AbstractContextualStorageManager.java:119) ~[classes:?]
	at java.util.Optional.ifPresent(Optional.java:183) ~[?:?]
	at com.vaadin.cdi.context.AbstractContextualStorageManager.decouple(AbstractContextualStorageManager.java:119) ~[classes:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177) ~[?:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497) ~[?:?]
	at com.vaadin.cdi.context.AbstractContextualStorageManager.destroy(AbstractContextualStorageManager.java:97) ~[classes:?]
	at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
	at com.vaadin.cdi.context.RouteScopedContext$ContextualStorageManager.onAfterNavigation(RouteScopedContext.java:64) ~[vaadin-cdi-11.2.0.jar:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:85) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.injection.MethodInvocationStrategy$SimpleMethodInvocationStrategy.invoke(MethodInvocationStrategy.java:168) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124) ~[jakarta.enterprise.cdi-api-2.0.2.jar!/:?]
	at org.jboss.weld.util.Observers.notify(Observers.java:166) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:177) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:159) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.manager.BeanManagerImpl.fireEvent(BeanManagerImpl.java:632) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at org.jboss.weld.util.ForwardingBeanManager.fireEvent(ForwardingBeanManager.java:104) ~[weld-core-impl-3.1.3.Final.jar!/:3.1.3.Final]
	at com.vaadin.cdi.CdiVaadinServletService$UIEventListener.afterNavigation(CdiVaadinServletService.java:229) ~[vaadin-cdi-11.2.0.jar:?]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.lambda$fireAfterNavigationListeners$3(AbstractNavigationStateRenderer.java:405) ~[flow-server-2.2.1.jar:2.2.1]
	at java.util.ArrayList.forEach(ArrayList.java:1540) ~[?:?]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.fireAfterNavigationListeners(AbstractNavigationStateRenderer.java:405) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.router.internal.AbstractNavigationStateRenderer.handle(AbstractNavigationStateRenderer.java:267) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.router.Router.handleNavigation(Router.java:249) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.router.Router.navigate(Router.java:220) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.router.Router.navigate(Router.java:186) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.component.UI.navigate(UI.java:892) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.component.UI.navigate(UI.java:862) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.component.UI.navigate(UI.java:817) ~[flow-server-2.2.1.jar:2.2.1]
	at com.upsilon.webportfolio.business.composition.CompositionTilesDesign.lambda$0(CompositionTilesDesign.java:106) ~[classes:?]
	at java.util.Optional.ifPresent(Optional.java:183) ~[?:?]
	at com.upsilon.webportfolio.business.composition.CompositionTilesDesign.onClickSecurity(CompositionTilesDesign.java:105) ~[classes:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:566) ~[?:?]
	at com.vaadin.flow.server.communication.rpc.PublishedServerEventHandlerRpcHandler.invokeMethod(PublishedServerEventHandlerRpcHandler.java:200) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.rpc.PublishedServerEventHandlerRpcHandler.invokeMethod(PublishedServerEventHandlerRpcHandler.java:174) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.rpc.PublishedServerEventHandlerRpcHandler.invokeMethod(PublishedServerEventHandlerRpcHandler.java:137) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.rpc.PublishedServerEventHandlerRpcHandler.handleNode(PublishedServerEventHandlerRpcHandler.java:125) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.rpc.AbstractRpcInvocationHandler.handle(AbstractRpcInvocationHandler.java:64) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocationData(ServerRpcHandler.java:409) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.ServerRpcHandler.lambda$handleInvocations$1(ServerRpcHandler.java:390) ~[flow-server-2.2.1.jar:2.2.1]
	at java.util.ArrayList.forEach(ArrayList.java:1540) ~[?:?]
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:390) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:317) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:89) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:40) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1545) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:247) ~[flow-server-2.2.1.jar:2.2.1]
	at com.vaadin.cdi.CdiVaadinServlet.service(CdiVaadinServlet.java:67) ~[vaadin-cdi-11.2.0.jar:?]
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:590) ~[jboss-servlet-api_4.0_spec-2.0.0.Final.jar!/:2.0.0.Final]
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.websockets.jsr.JsrWebSocketFilter.doFilter(JsrWebSocketFilter.java:173) ~[undertow-websockets-jsr-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:52) ~[opentracing-jaxrs2-0.4.1.jar!/:?]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at com.upsilon.webportfolio.security.SecurityRequestFilter.doFilter(SecurityRequestFilter.java:78) ~[classes:?]
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78) ~[?:?]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.RedirectDirHandler.handleRequest(RedirectDirHandler.java:68) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61) ~[?:?]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at org.wildfly.extension.undertow.security.jaspi.JASPICSecureResponseHandler.handleRequest(JASPICSecureResponseHandler.java:48) ~[?:?]
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68) ~[?:?]
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43) ~[undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:269) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:78) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:133) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:130) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43) ~[undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105) ~[?:?]
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541) ~[?:?]
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541) ~[?:?]
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541) ~[?:?]
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1541) ~[?:?]
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:249) [undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:78) [undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:99) [undertow-servlet-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:376) [undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830) [undertow-core-2.0.30.Final.jar!/:2.0.30.Final]
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35) [jboss-threads-2.3.3.Final.jar!/:2.3.3.Final]
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1982) [jboss-threads-2.3.3.Final.jar!/:2.3.3.Final]
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1486) [jboss-threads-2.3.3.Final.jar!/:2.3.3.Final]
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1377) [jboss-threads-2.3.3.Final.jar!/:2.3.3.Final]
	at java.lang.Thread.run(Thread.java:834) [?:?]
```

**Comment by roeltje25** [comment]
It seems in my case the first time the node is removed the StateNode.setParent method skips execution because hasDeta he'd is true.... I cannot understand yet why this is the case..... It will need some more debugging. Does anyone know how this may happen?

**Comment by roeltje25** [comment]
I have a test case that demonstrates the problem.... it took me some time. It is when one removes a parent in an overridden onDetach method. I cannot guess if this is a bug or forbidden usage...

```
package com.upsilon.webportfolio.utils;

import org.junit.Assert;
import org.junit.Test;

import com.vaadin.flow.component.DetachEvent;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.UI;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.dom.Element;

public class Tests
{
    @Tag("other")
    public static class OtherComponent extends Div {
        
    }

    @Test
    public void removeFromParent() {
        OtherComponent otherComponent = new OtherComponent( );
        UI ui = new UI( );
        Div parent = new Div( ) {
            @Override
            protected void onDetach(DetachEvent detachEvent)
            {
                otherComponent.getElement().removeFromParent( );
                super.onDetach( detachEvent );
            }
        };
        ui.add( parent );
        parent.getElement().appendChild( otherComponent.getElement() );
        Assert.assertTrue(otherComponent.getParent().isPresent());
        Assert.assertEquals(parent, otherComponent.getParent().get());
        parent.getElement().removeFromParent( );
        Assert.assertFalse(parent.getParent().isPresent( ));
        otherComponent.getElement().removeFromParent();
        Assert.assertFalse(otherComponent.getParent().isPresent( ));
    }

}
```

**Comment by denis-anisimov** [comment]
The test is failing because of IAE:
```
java.lang.IllegalArgumentException: Trying to detach an element from parent that does not have it.
	at com.vaadin.flow.dom.impl.AbstractNodeStateProvider.removeChild(AbstractNodeStateProvider.java:123)
	at com.vaadin.flow.dom.Node.removeChild(Node.java:376)
	at com.vaadin.flow.dom.Element.removeFromParent(Element.java:514)
```
on the line : `otherComponent.getElement().removeFromParent()`.

And this is absolutely expected :  `onDetach` of the parent removes the `otherElement` from parent : ` otherComponent.getElement().removeFromParent( )`.

`It is when one removes a parent in an overridden onDetach method`.
There is no overridden `onDetach`  method. The  default `onDetach` method doesn't do anything: it's no-op.
So there is no any need to call `super.onDetach( detachEvent )` and effectively it's the same as adding a detach listener.
There is no any need to use components in this test : as I said `onDetach`  doesn't produce any side effects because super impl doesn't do anything.

I don't see an issue here.
Closing the ticket since no steps to reproduce which allows to reproduce the ticket are provided.
Apparently the issue is in your code.

**Comment by roeltje25** [comment]
Dear Denis and Vaadin Team,

I am a bit taken aback by the pushback from my bug reports. It seems that either my reports are very difficult to understand or community input is not very welcome.

In the test case that I added you can clearly see an overridden onDetach method (if the parent method doesn't do anything that does not mean it is not overridden). Indeed I removed the parent in the onDetach method. In our code there was a reason for doing this, as the parent was not removed if we did not do it explicitly. Normally when removeFromParent is called the parent is set to null afterwards. removeFromParent can then be called an arbitrary amount of times again without issue because the null is detected and the method returns. However, when removing the parent in the onDetach method the parent is not set to null and a subsequent call to removeFromParent fails.

This particular issue only happens when using components (because of the onDetach method. had I used a detachllistener I would also have needed to use a Component or implement a DetachNotifier myself). 

I found it a bit strange that you claim I did not give steps to reproduce the ticket when clearly the testcase fails and as such contains clear steps to reproduce the issue.

Now, of course, it could be that this behavior is expected. However, neither the onDetach method, the DetachListener, the DetachEvent, nor the DetachNotifier class, nor the vaadin documentation mention that one cannot remove parents during the detach event firing. So, I would at least expect some clarification there.

In summary, please handle your community with a bit of care, because this does not invite me to delve into (possible) issues in vaadin in the future.

I would reopen the issue if I could.... but I do not have the rights

**Comment by denis-anisimov** [comment]
I can reopen the issue if you provide the steps to reproduce.

As I mentioned neither of provided tests work.

In the latest test you have the line `Assert.assertFalse(otherComponent.getParent().isPresent( ));` which I expect to be executed because otherwise I don't understand why it's there.

As I said in my last comment : The test is failing because of IAE.
The IAE is thrown way before the last line of the test.

So yes, I don't understand you: the test should fail because of assertion. It fails because of IAE.

>It seems that either my reports are very difficult to understand

Yes, It's impossible to understand when test says:

>Node.removeFromParent does not set parent of child node to null

but the real result in the steps is totally different.
Also this comment shows that the original test doesn't work: https://github.com/vaadin/flow/issues/8586#issuecomment-647367918

So two attempts to provide the information failed and I don't see reasons why to keep incomplete and at the moment invalid bug opened.
Not sure why it happens with your tickets, sorry :( We have a plenty good ticket with exact steps to reproduce / proper description with clear expected and actual result. 

As I said: please provide the steps to reproduce which really reproduce the problem and the expected result +actual result can be clearly seen or described in details.

**Comment by denis-anisimov** [comment]
Also please note that you may (and this is very desired) use this project https://github.com/vaadin/skeleton-starter-flow
as a base for the test case.
You may checkout, modify it (e.g. copy paste your test code there) and either zip it or make a Git repo whose link may be pasted here in the comments.
That allows to eliminate completely undesired unexpected things which may work in your specific code base in one way and in other way when we try to reproduce the issue.

**Comment by roeltje25** [comment]
Dear Denis,

The test has an assertFalse which should indeed run. I expect it to run. Indeed I do not expect the exception to be thrown. This is why one has test cases. This test case describes my problem exactly. My original title for the ticket may have been a bit to general. But I have explained better in later comments and provided (the final test case)  to illustrate the problem.

It is perfectly possible that a testcase throws an exception. This indicates that the test fails and the expected behavior is not present in the system. Or to use easy bullets:
1) I expect a particular behavior for which I know there is a problem
2) I create a test that should pass if the problem is not there and the behavior is there
3) it does not pass -> ergo it demonstrates there is a problem.

In effect the issue is that there is an IAE in the exact case described by the test case. I could make a skeleton project demonstrating the problem as well, but you already mentioned that the test case is failing for you as well, so is this necessary?

Again, It may very well be that no java code changes are needed, but in that case a clarification of some sort that one should never change the component tree during detach in the documentation would be needed. If that is the case, let me know and I will try to create a ticket there.

**Comment by denis-anisimov** [comment]
Dear Roel,

in the comment here I've already said:

IAE is expected.
If this is  the matter of the issue then yes, this is not a bug.

`parent` is removed from its "parent" : UI instance.
When it's removed the detach listener is called.
Detach listener calls `otherComponent.getElement().removeFromParent( )`.

Which means that `otherComponent.getElement().removeFromParent()` which  is called after that throws IAE.

I really don't know what I may add here: I consider this  absolutely clear and obvious.

Also the original code contains `e.getParent()` comparison to `null` so I expect that steps will lead to something that
confirms that `e.getParent()` will be `null`.
The provided test now says that it throws an exception.
I'm totally lost here: I don't see how these two things can replace each other. 

Apparently  I have serious problems with understanding so I would like to be aside of this.

But let's return back to the ticket:
- this is not a bug since I don't see this is as a bug. The explanation is above.
- So I  will keep the ticket closed.
- You are able to create another ticket . Since I'm not able to understand such tickets I will not touch it and someone who has this abilities will be able to handle this.

**Comment by roeltje25** [comment]
Dear Denis,

Although I was taken aback by the pushback I do value your input. It seems we are having difficulty understanding eachother and I hope this does not leave either of us with hard feelings. I may open a new bug with a clean slate and as much of a complete description as possible.

thank you and with kind regards,

Roel

**Comment by denis-anisimov** [comment]
Thank you very much !
👍

**Comment by roeltje25** [comment]
@denis-anisimov 

I created a new issue and would *value your input*, at least on my explanation and wording. I would like to create tickets that are good enough for you guys to understand and work with.
